cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +242,356 @@ impl UpdateConfig for Api {
-                x509
-            } else {
-                // Load the issuer and use it to sign the CSR.
+        if let Some(ref pem) = issuer_pem {
+            x509.extend_from_slice(pem)
+        }
 
-                let issuer_path = aziot_certd_config::util::get_path(
-                    &api.homedir_path,
-                    &api.preloaded_certs,
-                    issuer_id,
-                    true,
+        Ok(x509)
+    }
+    else {
+        let cert_options = cert_options
+            .ok_or_else(||
+                Error::invalid_parameter(
+                    "issuer",
+                    "issuer is required for locally-issued certs"
                 )
-                .map_err(|err| Error::Internal(InternalError::GetPath(err)))?;
-                let issuer_x509_pem = load_inner(&issuer_path)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?
-                    .ok_or_else(|| Error::invalid_parameter("issuer.certId", "not found"))?;
-                let issuer_x509 = openssl::x509::X509::stack_from_pem(&issuer_x509_pem)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-                let issuer_x509 = issuer_x509
-                    .get(0)
-                    .ok_or_else(|| Error::invalid_parameter("issuer.certId", "invalid issuer"))?;
-
-                x509.set_issuer_name(issuer_x509.subject_name())
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+            )?;
 
-                // Cap x509.not_after to issuer_x509.not_after
-                let issuer_not_after = issuer_x509.not_after();
-                let not_after = if issuer_not_after < not_after {
-                    issuer_not_after
-                } else {
-                    &not_after
-                };
-                x509.set_not_after(not_after)
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-                x509.sign(&issuer_private_key, openssl::hash::MessageDigest::sha256())
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
-
-                let x509 = x509.build();
+        match &cert_options.method {
+            CertIssuanceMethod::SelfSigned => {
+                let api_ = std::cell::RefCell::new(api);
+                let pk = api_.borrow().key_client.load_key_pair(id)?;
+                let api = api_.into_inner();
 
-                let mut x509 = x509
-                    .to_pem()
-                    .map_err(|err| Error::Internal(InternalError::CreateCert(Box::new(err))))?;
+                create_cert_inner(api, id, csr, Some((id, &CString::new(pk.0)?)))
+                    .await
+            },
+            CertIssuanceMethod::LocalCa => {
+                let CertificateWithPrivateKey { cert, pk } = api.cert_issuance.local_ca.clone()
```

Comment:
No reason to `clone()` it. Only `pk.0` is moved, so just `clone()` that as part of line 375.